pkg/azuredx/azureauth/auth.go

```diff
@@ -0,0 +1,133 @@
+// Package azureauth provides authorization utilities for the Azure cloud.
+package azureauth
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx/models"
+)
+
+var onBehalfOfLatencySeconds = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace: "grafana",
+	Subsystem: "plugin_adx",
+	Name:      "obo_latency_seconds",
+	Help:      "On-behalf-of token exchange duration.",
+	Buckets:   []float64{0.01, 0.05, 0.10, 0.20, 0.40, 1.00},
+}, []string{"http_status"})
+
+func init() {
+	prometheus.MustRegister(onBehalfOfLatencySeconds)
+}
+
+// ServiceCredentials provides authorization for cloud service usage.
+type ServiceCredentials struct {
+	models.DatasourceSettings
+	// PostForm is an http.Client method.
+	PostForm func(url string, data url.Values) (*http.Response, error)
+}
+
+// QueryDataAuthorization returns an HTTP Authorization-header value which
+// represents the respective request.
+// ⚠️ A zero string must cause appliance of the service account instead.
+//
+// TODO(pascaldekloe): Merge the service principal appliance in here. The entire
+//	authorization decission must be centralised for security reasons alone.
+//	A fully progmatic approach also bypasses the SDK's incorrect use of
+//	http.RoundTripper.
+func (c *ServiceCredentials) QueryDataAuthorization(req *backend.QueryDataRequest) (string, error) {
+	if !c.OnBehalfOf {
+		return "", nil
+	}
+
+	user := req.PluginContext.User // do nil-check once for all
+	if user == nil {
+		backend.Logger.Debug("using service principal for non-user request despite on-behalf-of configuration")
+		return "", nil
+	}
+
+	// Azure requires an ID token (instead of an access token) for the exchange.
+	userToken, ok := req.Headers["X-ID-Token"]
+	if !ok {
+		if _, ok := req.Headers["Authorization"]; !ok {
+			// Grafana permits a mixture of OAuth2 and system accounts.
+			backend.Logger.Warn("user credentials unavailable for data request; continue with service principal despite on-behalf-of configuration", "user", user.Login)
```

Comment:
Did you understand the comment above @andresmgot? It is easy to have the plugin fail on system accounts when the on-behalf-of switch is enabled. Some customers may manage datasources and panels with system accounts, and use OAuth2 accounts for the actual day-to-day monitoring. They would be stuck if we treat this like an error. Your call…

Reply:
Okay, yes as I say I am not too familiar with the Azure authentication model. Would this mean that a user could strip their user information from a query and impersonate the service principal?

If that's a yes and it could be a security issue it's probably better for the admin to create two datasources, one for users and another one for service accounts (and fail here). If not, this is fine. Thoughts?

Reply:
Correct @andresmgot. Since the test query (from the "Save & test" button on the configuration panel) also uses the service principal, regardless of the on-behalf-of checkbox state, we should be fine in terms of predictability.

Just to be clear, we do require the ID token to be present for all users whom login with OAuth2. It is just the service accounts—which must be explicitly created—that bypass the user propagation.